File: include/PSProjectFileData.java
package edu.uwstout.berriers.PSHelper.Model;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Stack;

import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamReader;

import com.trolltech.qt.core.QSettings;

public class PSProjectFileData extends PSXMLReader implements PSStatusDescribable {

	private File mPSProjectFile;
	private Stack<File> mPathStack;
	private String mPSVersion;
	
	// Project Chunks
	private ArrayList<PSChunkData> mChunks;
	private int mActiveChunk;

	public PSProjectFileData(File pPSProjectFile, QSettings settings) throws IOException {
		// Clear out any old chunks by re-initializing the array
		mChunks = new ArrayList<PSChunkData>();
		mActiveChunk = -1;

		if(pPSProjectFile != null) {
			mPSProjectFile = pPSProjectFile;
			mPSProjectFile = mPSProjectFile.getCanonicalFile();
			parseProjectFile();
			mActiveChunk = 0;
		}
	}

	public File getPSProjectFile() { return mPSProjectFile; }
	public String getPSVersion() { return mPSVersion; }

	public boolean parseProjectFile() throws IOException {
		// Sanity Checks
		if(mPSProjectFile == null || !mPSProjectFile.exists() || !mPSProjectFile.isFile()) {
			return false;
		}
			
	    XMLStreamReader reader = null;
	    mPathStack = new Stack<File>();
	    mPathStack.push(mPSProjectFile);
	    
	    try {
			// Get the starting XML stream
			InputStream lXMLFileStream = getXMLStreamFromFile(mPSProjectFile);
	    	reader = XML_FACTORY.createXMLStreamReader(lXMLFileStream);
	    	
	    	// Loop over all element tags
	        while (reader.hasNext()) {
	        	reader.next();
	            if(reader.isStartElement())
	            {
                    switch (reader.getLocalName()) {
                    
                    	// The document tag contains the PS XML version and possibly a path
                        case "document":
                        	// Extract PhotoScan file version
                        	mPSVersion = reader.getAttributeValue(null, "version");
                        	
                        	// Attempt to descend
                        	XMLStreamReader oldReader = reader;
                        	reader = explodeTag(reader, mPathStack);
                        	if(oldReader != reader) { oldReader.close(); }
                        break;
                    
                        // The chunks tag is an array of chunk tags
                        case "chunks":
                        	readElementArray(reader, "chunks", "chunk");
                        break;	                        
                    }
            	}	            
            }
	    } catch (XMLStreamException e) {
	    	throw new IOException("XML parsing error.", e);
	    } finally {
			try { if(reader != null) reader.close(); }
			catch (XMLStreamException e) {}
	    }
		
		return true;
	}
	
	@Override
	public void processArrayElement(XMLStreamReader reader, String elementName) {
		switch(elementName) {
			case "chunk":
				// Process this chunk tag
				PSChunkData lNewChunk = null;				
				try {
					lNewChunk = new PSChunkData(mPSProjectFile, reader, mPathStack);
				} catch (IOException e) {
					System.err.println("Error parsing chunk");
					e.printStackTrace();
					return;
				}
				
				// Save the results
				if(lNewChunk != null) {
					mChunks.add(lNewChunk);
				}
			break;
				
			default: super.processArrayElement(reader, elementName); break;
		}
	}

	public int getChunkCount() {
		return mChunks.size();
	}
	
	public int getActiveChunkIndex() {
		return mActiveChunk;
	}

	public PSChunkData getActiveChunk() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk);
		}

		return null;
	}

	public PSChunkData getChunk(int index) {
		if(index >= 0 && index < mChunks.size()) {
			return mChunks.get(index);
		}

		return null;
	}

	public File getModelArchiveFile() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getModelArchiveFile();
		}

		return null;
	}
	
	public PSModelData getModelData() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getModelData();
		}
		
		return null;
	}
	
	@Override
	public String describeImageAlignPhase() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).describeImageAlignPhase();
		}

		return "N/A";
	}

	@Override
	public byte getAlignPhaseStatus() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getAlignPhaseStatus();			
		}
		
		return 0;
	}

	@Override
	public String describeDenseCloudPhase() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).describeDenseCloudPhase();			
		}
		
		return "N/A";
	}

	@Override
	public int getDenseCloudDepthImages() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getDenseCloudDepthImages();			
		}
		
		return 0;
	}
	
	@Override
	public byte getDenseCloudPhaseStatus() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getDenseCloudPhaseStatus();			
		}
		
		return 0;		
	}

	@Override
	public String describeModelGenPhase() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).describeModelGenPhase();			
		}
		
		return "N/A";
	}

	@Override
	public byte getModelGenPhaseStatus() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getModelGenPhaseStatus();			
		}
		
		return 0;
	}

	@Override
	public long getModelFaceCount() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getModelFaceCount();			
		}
		
		return 0;
	}

	@Override
	public long getModelVertexCount() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getModelVertexCount();			
		}
		
		return 0;
	}

	@Override
	public String describeTextureGenPhase() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).describeTextureGenPhase();			
		}
		
		return "N/A";
	}

	@Override
	public byte getTextureGenPhaseStatus() {
		if(mActiveChunk >= 0 && mActiveChunk < mChunks.size()) {
			return mChunks.get(mActiveChunk).getTextureGenPhaseStatus();			
		}
		
		return 0;
	}
}

File: src/Model/PSProjectFileData.cpp
#include "PSProjectFileData.h"

#include <climits>
using namespace std;

#include <QFile>
#include <QSettings>
#include <QXmlStreamReader>

//#include "PSChunkData.h"
//#include "PSModelData.h"

PSProjectFileData::PSProjectFileData(QFile*& pPSProjectFile) { //}, QSettings*& settings) {
    // Clear out any old chunks by re-initializing the array
    mActiveChunk = UINT_MAX;

    if(pPSProjectFile != NULL) {
        mPSProjectFile = pPSProjectFile;
//        mPSProjectFile = mPSProjectFile->getCanonicalFile();
        parseProjectFile();
        mActiveChunk = 0;
    }
}

PSProjectFileData::~PSProjectFileData() {
    // Delete all chunk data
    for(unsigned int i=0; i < (unsigned int)mChunks.size(); i++) {
        delete mChunks[i];
    }

    // Clean up the file stack
    while(!mPathStack.empty()) {
        delete mPathStack.pop();
    }
}

QFile* PSProjectFileData::getPSProjectFile() { return mPSProjectFile; }
QString PSProjectFileData::getPSVersion() { return mPSVersion; }

bool PSProjectFileData::parseProjectFile() {
    // Sanity Checks
    if(mPSProjectFile == NULL || !mPSProjectFile->exists()) {
        return false;
    }

    // Initialize file stack
    QXmlStreamReader* reader = NULL;
    mPathStack.push(mPSProjectFile);

    try {
        // Get the starting XML stream
        reader = getXMLStreamFromFile(mPSProjectFile);

        // Loop over all element tags
        while (!reader->atEnd()) {
            reader->readNext();
            if(reader->isStartElement())
            {
                // The document tag contains the PS XML version and possibly a path
                if (reader->name() == "document") {
                    // Extract PhotoScan file version
                    mPSVersion = reader->attributes().value(NULL, "version").toString();

                    // Attempt to descend
                    QXmlStreamReader* oldReader = reader;
                    reader = explodeTag(reader, mPathStack);
                    if(oldReader != reader) {
                        delete oldReader;
                    }
                }

                // The chunks tag is an array of chunk tags
                else if (reader->name() == "chunks") {
                    readElementArray(reader, "chunks", "chunk");
                }
            }
        }
    } catch (...) {
        // Log error
        qWarning("XML Parsing encountered an error\n");

        // Finish and return failure
        delete reader;
        return false;
    }

    // Finish and return success
    delete reader;
    return true;
}

void PSProjectFileData::processArrayElement(QXmlStreamReader* reader, QString elementName) {
    if (elementName == "chunk") {
        // Process this chunk tag
        PSChunkData* lNewChunk = new PSChunkData(mPSProjectFile, reader, mPathStack);

        // Save the results
        if(lNewChunk != NULL) {
            mChunks.push_back(lNewChunk);
        }
    } else {
        PSXMLReader::processArrayElement(reader, elementName);
    }
}

unsigned int PSProjectFileData::getChunkCount() {
    return (unsigned int)mChunks.size();
}

unsigned int PSProjectFileData::getActiveChunkIndex() {
    return mActiveChunk;
}

PSChunkData* PSProjectFileData::getActiveChunk() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk];
    }

    return NULL;
}

PSChunkData* PSProjectFileData::getChunk(unsigned int index) {
    if(index < (unsigned int)mChunks.size()) {
        return mChunks[index];
    }

    return NULL;
}

QFile* PSProjectFileData::getModelArchiveFile() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getModelArchiveFile();
    }

    return NULL;
}

PSModelData* PSProjectFileData::getModelData() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getModelData();
    }

    return NULL;
}

QString PSProjectFileData::describeImageAlignPhase() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->describeImageAlignPhase();
    }

    return QString("N/A");
}

char PSProjectFileData::getAlignPhaseStatus() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getAlignPhaseStatus();
    }

    return 0;
}

QString PSProjectFileData::describeDenseCloudPhase() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->describeDenseCloudPhase();
    }

    return QString("N/A");
}

int PSProjectFileData::getDenseCloudDepthImages() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getDenseCloudDepthImages();
    }

    return 0;
}

char PSProjectFileData::getDenseCloudPhaseStatus() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getDenseCloudPhaseStatus();
    }

    return 0;
}

QString PSProjectFileData::describeModelGenPhase() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->describeModelGenPhase();
    }

    return QString("N/A");
}

char PSProjectFileData::getModelGenPhaseStatus() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getModelGenPhaseStatus();
    }

    return 0;
}

long PSProjectFileData::getModelFaceCount() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getModelFaceCount();
    }

    return 0;
}

long PSProjectFileData::getModelVertexCount() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getModelVertexCount();
    }

    return 0;
}

QString PSProjectFileData::describeTextureGenPhase() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->describeTextureGenPhase();
    }

    return QString("N/A");
}

char PSProjectFileData::getTextureGenPhaseStatus() {
    if(mActiveChunk < (unsigned int)mChunks.size()) {
        return mChunks[mActiveChunk]->getTextureGenPhaseStatus();
    }

    return 0;
}
